### PR TITLE
Handle project exit conditions

### PR DESCRIPTION
**Pull request by jcapona**
| Status  | Ticket/Issue |
| :---: | :--: |
| Ready/Hold | [Ticket](https://pi-top.atlassian.net/browse/OS-1273) |

#### Main changes
Handle project exit conditions:
- For now, the supported options that can be set in a configuration file using the `exit_condition` entry (eg: [here](https://github.com/pi-top/pi-top-4-Miniscreen/blob/fba51153e5e58df8bd938f0165802cb29e033da3/pt_miniscreen/demo_projects/pi_top_4/pong_game/project.cfg#L4)):

1. `POWER_BUTTON_PRESS`: press the pi-top[4] power button. This is now the default exit condition.
2. `HOLD_X`: press the 'X' (Cancel) button of the miniscreen for at least 3 seconds.
3. In the case an invalid or empty string is provided, the project will run until the provided command naturally stops.

- Added a new `Project` class that's in charge of running/waiting/stopping projects. This class is instantiated by a `ProjectPage` whenever the user decides to start a project.

- Updated ProjectPage text when project is started/running/stopped.

- Added exit conditions to pi-top[4] demos.

#### Screenshots (feature, test output, profiling, dev tools etc)

[insert screenshots here]

#### Other notes (e.g. implementation quirks, edge cases, questions / issues)
-

#### Manual testing tips
- Install artifacts from https://github.com/pi-top/pi-top-Python-SDK/pull/589 and https://github.com/pi-top/pi-topd/pull/175
- Install artifacts from this PR
- Restart `pt-miniscreen` service with `sudo systemctl restart pt-miniscreen`
- Once the app restarts, go to the Projects menu, access to the 'pi-top[4] Demos' page. 
- The pong game now has a 'power button press' exit condition, so once it starts, press the power button and the project should close.
- The 'hello world' demo now has a 'hold x' exit condition, once it starts hold the X button for more than 3 seconds and the project should close.

#### Tag anyone who definitely needs to review or help
-

### CODE REVIEW

**Review by olivierwilkinson — COMMENTED** [review]
The exit conditions themselves work great, there are a few things I noticed while checking them though:
- The 'Hold X' exit not only exits you out of the project page, it also exits you from your directory if you hold too long; this is probably because we go up a directory on cancel button release. Edit: maybe we should go back to the stopwatch idea, then it doesn't close the project until the user lets go)
- After exiting, the screen that says "Running {project}" appears for a flash before animating away. Preferably this wouldn't be shown at all and you would land on the projects directory again.
- There is no indication which exit condition is being used for which project, this should be added to the "Running {project}" page I think.

There are some other things that are semi-related but shouldn't be addressed in this PR:
- the projects menu doesn't have any icons in the gutter for entering a directory or starting a project
- When starting the project preferably the screen that says "Running {project}" wouldn't animate in, it would feel more responsive if it took over the screen immediately.
- A more minor point is that the screen should probably say "Starting {project}..."

Overall I think it's really taking shape though, I think it's great in general! The above is almost all stuff that is fairly minor and/or easy to resolve. 👍

**Comment by olivierwilkinson** [comment]
The new changes are good, I like the new starting screen much better. There are a couple things still to do I think:
- when the starting screen goes blank there is a 2s pause before the project starts now, before it was immediate.
- The blank screen after closing the project should say "Stopping {project}..."
- I've [raised a PR for pushing and popping from the Stack without animating](https://github.com/pi-top/pi-top-4-Miniscreen/pull/296). I think if you can review that (It's pretty simple) and then use it when pushing and popping the start / stop screen it would improve the feel of this a lot.

File: pt_miniscreen/demo_projects/pi_top_4/pong_game/project.cfg [hunk 1]
```diff
@@ -1,3 +1,4 @@
 [project]
 title=Pong Game
 start=python3 -B pong.py
+exit_condition=POWER_BUTTON_PRESS
```
[the remark this answers is not in the document]
Comment (olivierwilkinson) [reply pt_miniscreen/demo_projects/pi_top_4/pong_game/project.cfg:4]: Press, then if someone is holding because they can't remember how to turn off the project it exits before the pi-top powers down. It will still work with a flick as well :P. The PR description is a little out of date.

File: pt_miniscreen/pages/root/projects.py [hunk 2]
```diff
@@ -64,86 +68,190 @@ def from_file(cls, file):
...
-            on_stop()
+    @property
+    def displayed_text(self) -> str:
+        text = ""
+        if self.state == ProjectState.ERROR:
+            text = f"There was an error while running '{self.project_config.title}'..."
+        elif self.state == ProjectState.STOPPING:
+            text = f"Stopping '{self.project_config.title}'..."
+        elif self.state == ProjectState.STARTING:
+            text = f"Starting '{self.project_config.title}'..."
```
Comment (olivierwilkinson) [inline pt_miniscreen/pages/root/projects.py:205]: There should also be text for when the project is running since there is no guarantee that user projects take over the miniscreen.

```suggestion
        elif self.state == ProjectState.STARTING:
            text = f"Starting '{self.project_config.title}'..."
        elif self.state == ProjectState.RUNNING:
            text = f"Running '{self.project_config.title}'..."
```

File: pt_miniscreen/pages/root/projects.py [hunk 3]
```diff
@@ -199,6 +198,8 @@ def displayed_text(self) -> str:
         text = ""
         if self.state == ProjectState.ERROR:
```
Comment (olivierwilkinson) [inline pt_miniscreen/pages/root/projects.py]: I think these need to be:
```self.state["project_state"] == ProjectState.ERROR```

File: pt_miniscreen/pages/root/projects.py [hunk 4]
```diff
@@ -64,86 +68,190 @@ def from_file(cls, file):
...
+        super().__init__(**kwargs, initial_state={"project_state": ProjectState.IDLE})
 
-        self.process = None
         self.text = self.create_child(
-            MarqueeText,
-            text=f"Running '{self.project_config.title}'",
-            font_size=14,
+            Text,
+            text=self.displayed_text,
+            get_text=lambda: self.displayed_text,
```
Comment (olivierwilkinson) [inline pt_miniscreen/pages/root/projects.py:190]: Ah! This is why it feels a bit delayed: get_text has a default interval of 1s.

I think if you remove this line and set the text when state changes it will update immediately:

```python
    def on_state_change(self, previous_state):
        if self.state["project_state"] != previous_state["project_state"]:
            self.text.state.update({"text": self.displayed_text})
```

**Review by olivierwilkinson — APPROVED** [review]
Works great, really well done! 🙌

**Review by angusjfw — CHANGES_REQUESTED** [review]
- When a project ends naturally rather than using an exit condition, it displays the 'running X' message again before showing 'Stopping X'. Eg try the 'Hello!' demo and let it run to exit.

- Do we really need the 'Running X' message after already showing 'Starting X... do Y to exit'? It looks particularly silly when you use exit_condition=NONE...

- I found that if I try to run this demo project while running a simple `while True` loop in Further then starting fails with a message in the journal `error: Error starting project: Exception occurred in preexec_fn.`

- Maybe we should describe it as 'flick' power rather than press just to be clear that you shouldn't hold it long (in the message and the config option). I think this is what Mike was asking about. Jesse also seemed concerned by using the power button for this, though I think it works ok in practice.

**Comment by olivierwilkinson** [comment]
> - When a project ends naturally rather than using an exit condition, it displays the 'running X' message again for a few seconds before showing 'Stopping X'. Eg try the 'Hello!' demo and let it run to exit.

This is because of the implementation of hello.py, it has a sleep(5) after displaying the text using the miniscreen so the project is in fact still running.

> - Maybe instead of 'Running X' message we should say 'Started X'.  At the moment it's not clear why we say 'Running X' after already saying 'Starting X', they mean the same thing. 'X is running' could work too.

I'm also ok going with 'X is running', the difference between starting and running is subtle to be fair.

> - Maybe we should describe it as 'flick' power rather than press just to be clear that you shouldn't hold it long (in the message and the config option). I think this is what Mike was asking about.

@m-roberts is this what you meant? If so I'm sorry, I misunderstood!

I still think press is a better description of what the trigger is, and now that the feedback that the project is stopping is immediate I'm not worried about people accidentally turning their pi-top off (I think Jesse would agree with it in his hands).

Said another way, I can see users trying to flick but not triggering the exit condition, particularly kids as they may not press hard enough.

**Comment by jcapona** [comment]
> When a project ends naturally rather than using an exit condition, it displays the 'running X' message again for a few seconds before showing 'Stopping X'. Eg try the 'Hello!' demo and let it run to exit.

hey @angusjfw  - just pushed some changes that fix this issue by setting the project page text message to be empty whenever a project uses the miniscreen. When such a project naturally finishes, the screen goes to black for a second (while stopping and cleaning up), then shows the 'stopping project' message.

> Maybe instead of 'Running X' message we should say 'Started X'. At the moment it's not clear why we say 'Running X' after already saying 'Starting X', they mean the same thing. 'X is running' could work too.

makes sense - changed it to 'X is running'

> I found that if I try to run this demo project while running a simple while True loop in Further then starting fails with a message in the journal error: Error starting project: Exception occurred in preexec_fn. This doesn't happen if I run the same from a python console though.

I can replicate this but not really sure what's the issue there; I'll create a ticket to investigate and find a solution for this... I don't expect a lot of users to find this issue one though 

> Maybe we should describe it as 'flick' power rather than press just to be clear that you shouldn't hold it long (in the message and the config option). I think this is what Mike was asking about. Jesse also seemed concerned by using the power button for this, though I think it works ok in practice.

changed the code to use 'flick' instead of 'press'

**Comment by angusjfw** [comment]
Works great.

One final nitpick would be that `HOLD_X` and `POWER_BUTTON_FLICK` are inconsistent...

I think my preference would be action then target eg `HOLD_X` and `FLICK_POWER`... avoids debate about whether it's a POWER_BUTTON or a POWER_MOMENTARY_SWITCH 🙃
